File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists)==0:
            return None
        A=[]
        for i in range(len(lists)):
            if lists[i]:
                heappush(A,[lists[i].val,i])
        head=ListNode(0)
        curr=head
        while(A):
            temp=heappop(A)
            curr.next=lists[temp[1]]
            lists[temp[1]]=lists[temp[1]].next
            if lists[temp[1]]:
                heappush(A,[lists[temp[1]].val,temp[1]])
            curr=curr.next
        return head.next
            
            
        

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
    priority_queue<pair<int,ListNode*>,vector<pair<int,ListNode*>>,greater<pair<int,ListNode*>>>heap;    
     ListNode dummy;
        ListNode* curr=&dummy;
    for(auto x:lists)if(x!=nullptr)heap.push({x->val,x});
    while(!heap.empty())
    {
        pair<int,ListNode*>temp=heap.top();
        heap.pop();
        ListNode* Next=temp.second->next;
        curr->next=temp.second;
        curr=curr->next;
        if(Next!=nullptr)heap.push({Next->val,Next});
    }
    curr->next=nullptr;
    return dummy.next;
        
        
        
        
        
        
        
        
        
        
    }
};